boa/src/value/to_integer_or_infinity.rs

```diff
@@ -0,0 +1,99 @@
+use super::*;
+
+use std::convert::TryFrom;
+
+/// Represents the result of ToIntegerOrInfinity operation
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum IntegerOrInfinity {
+    Integer(isize),
+    Undefined,
```

Comment:
Why `Undefined`? It can only be (+ or -) infinity or an integer, right?

Reply:
Not necessarily, [ToNumber](https://tc39.es/ecma262/#sec-tonumber) returns `NaN` in case of `undefined` argument (that is - mostly when it is not specified at all) and it needs to be handled.

Reply:
> Not necessarily, [ToNumber](https://tc39.es/ecma262/#sec-tonumber) returns `NaN` in case of `undefined` argument (that is - mostly when it is not specified at all) and it needs to be handled.

But in the case that `ToNumber` returns `NaN`, and in step 2 of [`ToIntegerOrInfinity`](https://tc39.es/ecma262/#sec-tointegerorinfinity) it's converted to `0`, so ToIntegerOrInfinity(undefined) returns an integer `0`.
I don't see how it could return undefined, Am I missing something?

Reply:
Yeah, you're right. I forgot why I did that in the first place and jumped to conclusion too quickly.

The `Undefined` case has nothing to do with `ToIntegerOrInfinity` - I did add the special case here so that I can add [`as_relative_start` and `as_relative_end`](https://github.com/boa-dev/boa/pull/919/files#diff-907d77dea9f86d13f0ec081a2df726de54ad4e7ccbbb72c4edaf836a12bec639R16-R49) helpers and abstract-out the range calculation in `Array` methods that work on ranges (e.g. [`fill`](https://github.com/boa-dev/boa/pull/919/files#diff-2cb9ec2c93c8ac061e6eeb85d02353c854d2b26246901361aa61646bcbfdb186R882-R891)). Taking `fill` as an example, the algorithm in spec mentions (text in `[]` is my comment)

> Let relativeStart be ? ToIntegerOrInfinity(start). [this makes `undefined` == 0]
> (...)
> If end is undefined, let relativeEnd be len; else let relativeEnd be ? ToIntegerOrInfinity(end). [this makes `undefined` == `len`]

It is used to define the range being worked on. You can see the asymmetry there. If we move the `undefined` handling out of `ToNumber`/`ToIntegerOrInfinity` (i.e. make `ToIntegerOrInfinity` return `undefined` in case of `undefined`) to the call site, we can more or less make them symmetric and have the same calculation used in all of the implementations. That's what I did with the aforementioned helpers.

This might trade the complexity out of `array.rs` into some other module that will use `ToIntegerOrInfinity` in the future, but I think that the other cases (notably - calculating index instead of range) would work with another helper like `as_index`.